plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,476 @@
+
+    private final Repository repository;
+
+    private final StorageFacet storageFacet;
+
+    private final MavenFacet mavenFacet;
+
+    private final MavenPathParser mavenPathParser;
+
+    private final MetadataBuilder metadataBuilder;
+
+    private final MetadataUpdater metadataUpdater;
+
+    private final DocumentBuilderFactory documentBuilderFactory;
+
+    public Worker(final Repository repository,
+                  final boolean update,
+                  final String sql,
+                  final Map<String, Object> sqlParams)
+    {
+      this.sql = sql;
+      this.sqlParams = sqlParams;
+      this.repository = repository;
+      this.storageFacet = repository.facet(StorageFacet.class);
+      this.mavenFacet = repository.facet(MavenFacet.class);
+      this.mavenPathParser = mavenFacet.getMavenPathParser();
+      this.metadataBuilder = new MetadataBuilder();
+      this.metadataUpdater = new MetadataUpdater(update, repository);
+      this.documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    /**
+     * Method rebuilding metadata that performs the group level processing. It uses memory conservative "async" SQL
+     * approach, and calls {@link #rebuildMetadataInner(String, String, Set)} method as results are arriving.
+     */
+    public void rebuildMetadata(final ODatabaseDocumentTx db)
+    {
+      db.command(
+          new OSQLAsynchQuery<ODocument>(
+              sql,
+              new OCommandResultListener()
+              {
+                String currentGroupId = null;
+
+                @Override
+                public boolean result(Object iRecord) {
+                  final ODocument doc = (ODocument) iRecord;
+                  final String groupId = doc.field("groupId", OType.STRING);
+                  final String artifactId = doc.field("artifactId", OType.STRING);
+                  final Set<String> baseVersions = doc.field("baseVersions", OType.EMBEDDEDSET);
+
+                  final boolean groupChange = !Objects.equals(currentGroupId, groupId);
+                  if (groupChange) {
+                    if (currentGroupId != null) {
+                      rebuildMetadataExitGroup(currentGroupId);
+                    }
+                    currentGroupId = groupId;
+                    metadataBuilder.onEnterGroupId(groupId);
+                  }
+                  rebuildMetadataInner(groupId, artifactId, baseVersions);
+                  return true;
+                }
+
+                @Override
+                public void end() {
+                  if (currentGroupId != null) {
+                    rebuildMetadataExitGroup(currentGroupId);
+                  }
+                }
+              }
+          )
+      ).execute(sqlParams);
+    }
+
+    /**
+     * Process exits from group level, executed in isolation.
+     */
+    private void rebuildMetadataExitGroup(final String currentGroupId) {
+      isolate(new Runnable()
```

Comment:
So, something's not sitting right with the isolate method.  You're running a long, batch process that can't follow StorageTx's approach to transactions, but yet you need the helper methods build into StorageTx.  Essentially, StorageTx does two jobs - transaction control, and provides a suite of domain model manipulation tools (e.g. manipulating Components and Assets).

The isolate() method is necessary because of Orient DB's underlying addiction to thread-locals, and you're trying to use some of the functionality of StorageTx without being impacted by its assumption that it can control Orient.

Rather than pepper our code with this sort of suppression/hack stuff, I think we ought to tease apart the transaction control and domain-manipulation features of StorageTx - or perhaps, encapsulate the way that we're obtaining connections/transactions from Orient into a service that StorageTx also uses, to allow for this sort of isolation globally.

Anyways, let's not hold up this PR with this, but queue up another issue to deal with soon.